Feat/initial dev

RobertRosca opened the pull request:
The application serves as a "Zulip Write Only Proxy," providing a web API to interact with Zulip services. It allows clients to send messages, upload images, and list topics within specific streams. The application also manages client authentication and implements a JSON-based repository for data persistence.

## Software Choices

Project:

- **Poetry** for python project management.
- **Docker** for containers.

Architecture:

- **FastAPI**: as the web framework.
- **Pydantic**: data validation and parsing.
- **Uvicorn** as the web server.
- **Docker** containers for deployment.
- **Docker compose**/stacks for continuous no-downtime deployment.
  - Multiple replicas of the proxy service run at once.
  - **nginx** acts as load balancer in front of them.

## Description

The application (roughly) follows domain-driven design principles, creating a separation between different concerns and layers. DDD isn't really used in DA so I'll write a short description of the concepts.

### Domain Driven Design Summary

Domain-driven design is a software development methodology that aims to create a separation between different concerns and layers. It is based on the idea that the domain (the problem being solved) should be the main focus of the software, and that the code should be organized around it. The domain is divided into different layers, each with its own responsibilities and concerns.

The layers used in this project are:

- **Domain layer**: the core business logic and domain logic of the application. It contains only the logic related to solving the problem at hand, and is independent of the other layers.
- **Application/Service layer**: the layer that coordinates between the domain and data access layers. It contains the logic for creating, retrieving, and listing clients.
- **Infrastructure/Repository layer**: the layer that manages data access and storage. It contains the logic for interacting with the JSON-based file storage.
- **Presentation/API layer**: the layer that exposes the functionality through FastAPI endpoints. It connects the different layers and includes security measures like API key headers.

DDD is an 'onion' architecture with the domain at the core and everything else depending on the domain.

### Domain Layer

The domain layer in this case is a single file `models.py`. It contains models for the scoped client (a Zulip client scoped to a single stream), and an admin client (an internal client which has the ability to create new scoped clients).

In DDD the domain layer is at the center of the application, and is independent of the other layers. It should not import anything from the other layers, should not be aware of infrastructure concerns like persistence/storage, and should not be aware of presentation concerns like web APIs.

### Infrastructure/Repository Layer

The infrastructure layer in this case is a single file `repository.py`. For simplicity right now this is just a JSON-based file storage, but it could be replaced with a database or other storage solution in the future.

In DDD the infrastructure layer is responsible for data access and storage. It should be aware of domain concerns, since it returns domain objects, but should not be aware of the presentation layer.

### Application/Service Layer

The application layer in this case is a single file `services.py`. It contains methods for creating, retrieving, and listing clients from an instance of a repository.

In DDD the application layer is responsible for coordinating between the domain and data access layers. It should be aware of both the domain and the infrastructure layers, but should not be aware of the presentation layer.

### Presentation/API Layer

The presentation layer in this case is the `main.py` file and `cli.py`. It contains FastAPI endpoints and CLI commands for creating, retrieving, and listing clients. It also includes security measures like API key headers.

In DDD the presentation layer is responsible for exposing the functionality to users, in this case through a web API and through a CLI interface. It should be aware of the application/service layer, but should not be aware of the domain or infrastructure layers.

### Deployment

The production deployment involves multiple replicas of the service running in front of nginx. The nginx container acts as a load balancer, distributing requests between the replicas. This allows for continuous deployment with no downtime, since the replicas can be updated one at a time.

This is all defined in a docker compose file, but compose is **not** used for the deployment, instead docker stacks are used. Stacks are pretty much docker compose (they use compose file) but with some fancier features available, in this case the relevant features are in the `deploy` section of `docker-compose.yml`:

```yaml
    deploy:
      update_config:
        parallelism: 1
        delay: 10s
        failure_action: rollback
        order: start-first
        monitor: 5s
        max_failure_ratio: 0
```

This configuration means that when the stack is updated, it will start up a new instance of the container, and wait for it to be running successfully. If the update worked then the old instance is removed, if an update fail it will leave the original container running.

Since the replicas are updated one at a time, and since nginx is configured to distribute requests between them, this allows for continuous deployment with no downtime caused by the update process.

'Normal' docker compose deployments would pull in the image, stop the old container, and start the new one, which would cause downtime. If the new container fails to start then the old one is already stopped so the service would be down. Stacks allow for a more controlled update process.

## Todo

Required:

- ~~[ ] Simplify tests: A lot of tests are redundant, test scope should be limited to the responsibilities of a layer.~~ already wrote the tests, might as well keep them
- [x] Contribution documentation: `poe` commands, code style guide, commit style guide, formatting, type checking.
- [x] Fix `/me` endpoint for admin
- [x] Un-censor key for create endpoint :eyes: 
- [x] Before merge, change the image tag to `zulip-write-only-proxy:main` (or latest? I forget which tags the metadata extractor creates)

Optional:

- ~~[ ] Docs? Overkill since this is so simple, but also very easy to set up. Repo is public and this kind of stream-scoped client might be useful to others using Zulip bots so maybe docs are worth including.~~

RobertRosca committed:
feat(endpoint): add initial features

RobertRosca committed:
feat(endpoint): allow for image uploads

RobertRosca committed:
feat(endpoint): add very rudimentary authentication

RobertRosca committed:
chore: add docstrings, add cli as script

RobertRosca committed:
style: sort imports

RobertRosca committed:
refactor: rename classes/methods to maybe make more sense

RobertRosca committed:
feat: Zulip client setup in service, add message/upload methods to model

RobertRosca committed:
refactor: more renaming, proposal to client, token to key

RobertRosca committed:
fix(endpoint): use proper header auth

RobertRosca committed:
feat(cli): allow specifying client stream/topic in CLI

RobertRosca committed:
build(docker): add Dockerfile

RobertRosca committed:
feat: store clients/zuliprc in cwd for easy config/start

RobertRosca committed:
chore(docker): use default port 8000

RobertRosca committed:
feat: better exception handling

RobertRosca committed:
docs: add readme

RobertRosca committed:
chore: update Zulip client injection

RobertRosca committed:
fix: newline before image URI insertion

RobertRosca committed:
feat(endpoint): add upload image endpoint

RobertRosca committed:
chore(docker): use buildkit caching

RobertRosca committed:
chore(docker): add compose file for easier pull/up/down

RobertRosca committed:
build(deps): lower Python version to ^3.9

RobertRosca committed:
chore(deps): make compatible with Python 3.9

RobertRosca committed:
chore(docker): use .env to set port

RobertRosca committed:
docs: reference Zulip docs, drop response validation

RobertRosca committed:
feat(endpoint): allow topic to be set by API call

RobertRosca committed:
feat: add Admin client, use locking for client.json writes

RobertRosca committed:
style: more consistent imports

RobertRosca committed:
docs: update usage documentation

RobertRosca committed:
fix: oversight on admin check

RobertRosca committed:
feat(endpoint): add list topics endpoint

RobertRosca committed:
refactor: rename to plural

RobertRosca committed:
chore(deps): use pydantic 2.0 field validator decorator

RobertRosca committed:
style: formatting improvements

RobertRosca committed:
build(add): add pytest

RobertRosca committed:
fix: store files in config directory to deal with docker mount

RobertRosca committed:
refactor: remove redundant response model

RobertRosca committed:
fix: update response for invalid key

RobertRosca committed:
fix: deal with mix of client/admins in repo list

RobertRosca committed:
refactor: use factory to return same object instead of default

Default actually returns a deep copy of the object

RobertRosca committed:
test: add initial tests

RobertRosca committed:
feat: add bot to stream when client is created

RobertRosca committed:
feat: use poe task to serve

RobertRosca committed:
test: update tests for default factory ScopedClient._client

RobertRosca committed:
feat(endpoint): add `/me` endpoint to get details on what stream you use

RobertRosca committed:
feat(model): change key field to SecretStr

RobertRosca committed:
test: add more tests

RobertRosca committed:
build(add): add ruff for linting/analysis

RobertRosca committed:
build(add): add pyright, update lint command

RobertRosca committed:
refactor: use FastAPI Annotated style instead of kwarg style

RobertRosca committed:
fix: lint/type issues

RobertRosca committed:
build: switch to pytest-cov

RobertRosca committed:
fix: incorrect annotation

RobertRosca committed:
build: split up lint tasks

RobertRosca committed:
ci(test): create tests.yml

RobertRosca committed:
ci(lint): create lint.yml

RobertRosca committed:
ci(style): consistent action names

RobertRosca committed:
fix: test issue when no zuliprc file present

RobertRosca committed:
refactor: import module not funcs, add annotations

RobertRosca committed:
style: fix formatting

RobertRosca committed:
ci(test): upload test coverage

RobertRosca committed:
ci(docker): add ci to build and push docker image

RobertRosca committed:
chore(deploy): configure auto update and load balancer

RobertRosca committed:
chore(deploy): remove watchtower, use compose 'update_config'

RobertRosca committed:
feat(endpoint): add health endpoint

RobertRosca committed:
chore(deploy): change docker compose update config

RobertRosca committed:
docs: add deployment section

RobertRosca committed:
chore(deploy): add nginx healthcheck

RobertRosca committed:
chore(deploy): pin nginx version

RobertRosca committed:
ci(docker): do not build image on pr trigger

RobertRosca committed:
build(add): add commitizen

RobertRosca committed:
chore: add black task, add format task, set lint to check only

RobertRosca committed:
ci: continue on error for lint checks

RobertRosca committed:
build(deps): add black to lint group

RobertRosca committed:
ci: replace continue on error with if always

Continue on error marks workflow as passed even if step fails. Idea is to have all steps keep running even if one fails, but still mark the workflow as failed.

RobertRosca committed:
ci: rename lint black step, add check flag

RobertRosca committed:
style: set line length to 88

RobertRosca committed:
style: reformat

CammilleCC commented:
Discussed with Robert about the implementation details and my questions were nicely answered. The DDD approach looks good; the layers are nicely separated and the client models are quite abstracted. The tests also describe the behavior of the functions and the endpoints in different scenarios.

Looking forward to the other TODOs. For me, docs are not much needed as there is already a detailed README for the usage.

LGTM

RobertRosca committed:
feat(endpoint): deprecate `upload_image`, rename to `upload_file`

Done to keep in line with native Zulip API endpoints

`upload_image` deprecated, will be removed in next commit

RobertRosca committed:
feat(endpoint): remove `upload_image`

Removes deprecated `upload_image` endpoint

RobertRosca committed:
build: add makefile for easier up/down with stacks

RobertRosca committed:
docs(deploy): update documentation w/ make commands

RobertRosca committed:
test: replace some references to image with generic file

RobertRosca committed:
build(deps): update pyright

RobertRosca committed:
ci: create reusable composite action for project setup

RobertRosca committed:
ci: use reusable actions for lint and test

RobertRosca committed:
ci: do not build on all branches, set workflow call/dispatch

RobertRosca committed:
ci: build image after successful tests

RobertRosca committed:
feat(endpoint): rename `/message` to `/send_message`, deprecate `/message`

RobertRosca committed:
feat(endpoint): add `/update_message` endpoint and model methods

RobertRosca committed:
fix(deploy): add `git pull` to make up, fix compose typo

RobertRosca committed:
fix(deploy): remove quotes around nginx port

RobertRosca committed:
chore(deploy): erorr on missing port env var

RobertRosca committed:
build(deploy): parse compose config with sed before pipe to stack deploy

various interoperability problems between compose
spec and stack spec, see https://github.com/docker/cli/issues/1073

RobertRosca committed:
ci(deploy): set push to true

RobertRosca committed:
fix(deploy): set head sha on pr event

RobertRosca committed:
fix(deploy): use pr number instead of branch name

RobertRosca committed:
chore(deploy): update image url to pull from european-xfel

RobertRosca committed:
chore(deploy): remove `replicas: 2` since `order: start-first` achieves desired effect

RobertRosca committed:
feat(endpoint): remove deprecated '/message' endpoint

old '/message' endpoint no longer works

RobertRosca committed:
feat(endpoint): make message contents a body parameter

message contents no longer query, api calls need to be changed see
https://github.com/European-XFEL/DAMNIT/pull/76#pullrequestreview-1612583411

RobertRosca committed:
test: update tests after removing deprecated endpoint

RobertRosca committed:
feat: remane `list_topics` to `get_stream_topics`

aligns closer to Zulip API

endpoint renamed

RobertRosca committed:
feat: `get_stream_topics` return zulip response instead of custom error

RobertRosca committed:
docs: document poe commands

RobertRosca committed:
docs: simplify readme, update endpoints

RobertRosca committed:
build: bump pyright

RobertRosca committed:
fix(cli): run services setup before creating client

RobertRosca committed:
fix(models): do not log stream as `extra`

the `stream` dict contains a `msg` key which causes issues with stdlib logging

RobertRosca committed:
fix: rename `get_topics` to `get_stream_topics`

RobertRosca committed:
feat(endpoint): make topic/content conditionally required for update message

RobertRosca committed:
test: exclude cli module from coverage

RobertRosca committed:
build(deps): update pyright

RobertRosca committed:
fix: add `from __future__ import annotations`

RobertRosca committed:
test: remove python 3.9 from test matrix

RobertRosca committed:
build: bump python, fastapi, pydantic, ruff

RobertRosca committed:
test(endpoint): add tests for `/me` endpoint

RobertRosca committed:
fix(endpoint): fix `/me` endpoint response type missmatch

RobertRosca committed:
fix(endpoint): update mypy ignore categories

RobertRosca committed:
feat(endpoint): explicitly exclude `Client.key` from `/me` response

RobertRosca committed:
feat(endpoint): tag `/health` as `Admin`

RobertRosca committed:
fix(endpoint): return client key on `/create_client` call

perviously returned the obfuscated `SecretStr`

RobertRosca committed:
test(endpoint): test that `/create_client` returns exposed secret

RobertRosca committed:
ci: bump python version for lint workflow to 3.11

RobertRosca committed:
build: bump max python version to 3.13

RobertRosca committed:
build(deploy): use port 8080 as standard

RobertRosca committed:
build(deploy): allow setting image and tag via env vars

RobertRosca committed:
build(dev): add 'dev-docker' target to Makefile

RobertRosca committed:
build: use poetry-dynamic-versioning

RobertRosca committed:
feat(endpoint): include version info in health response

RobertRosca committed:
feat(logging): exclude '/health' endpoint calls from logs

RobertRosca committed:
docs: add todo section

RobertRosca committed:
docs: add client creation section

RobertRosca committed:
docs: update deployment section

RobertRosca committed:
docs: update development section

RobertRosca committed:
docs: replace exfldadev01 with localhost

RobertRosca committed:
style: format, fix mypy error

RobertRosca committed:
build: update lockfile

RobertRosca committed:
docs: add dependabot to todo list

RobertRosca committed:
docs: add debugging section

RobertRosca committed:
ci: add dynamic versioning plugin during setup

RobertRosca committed:
build(deploy): expose port 8080 in dockerfile

RobertRosca commented:
@CammilleCC @matheuscteo ready for review. No major changes since you both last looked at it.

RobertRosca committed:
fix(endpoint): import version info from module __init__ instead of from _version

_version might not exist depending on install method/context, __init__ imports handle this

RobertRosca committed:
docs: add build-image workflow test and client bot config to todo list

matheuscteo commented:
Thank you Robert, LGTM.

CammilleCC reviewed (approved):
Tested in production, thus LGTM
